### DISCUSSION TITLE
1.1.0

### DISCUSSION THREAD

**Discussion by mglaman** [mglaman/phpstan-drupal#253] [Announcements]
## Backwards compatibility warning

1.1.0 has introduced a breaking change in the configuration schema. See below for more information.

## What's Changed
* Test drupal container related type inference by @brambaud in https://github.com/mglaman/phpstan-drupal/pull/227
* BrowserTestBaseDefaultThemeRule false positives with UpdatePathTests by @mglaman in https://github.com/mglaman/phpstan-drupal/pull/235
* Update branch alias in composer.json by @jibran in https://github.com/mglaman/phpstan-drupal/pull/236
* Test EntityInterface return type by @brambaud in https://github.com/mglaman/phpstan-drupal/pull/237
* Allow magic methods for variables of type FieldItemListInterface by @eiriksm in https://github.com/mglaman/phpstan-drupal/pull/244
* Support entity query execute dynamic return type whether it is a count query or not by @brambaud in https://github.com/mglaman/phpstan-drupal/pull/241
* Support entity class return type extension from entity storage methods by @brambaud in https://github.com/mglaman/phpstan-drupal/pull/239
* EntityDataRepository for entity type data by @mglaman in https://github.com/mglaman/phpstan-drupal/pull/250
* Use entity storage class inferred from entity type by @mglaman in https://github.com/mglaman/phpstan-drupal/pull/251

**Full Changelog**: https://github.com/mglaman/phpstan-drupal/compare/1.0.3...1.1.0

## Configuration changes

The `drupal.entityTypeStorageMapping` allowed specifying what storage classes were used for different entity types. With #239, support was added to specify what entity class an entity type uses. With #250 the way entity type information is configured has been changed.

Previously:

```
parameters:
	drupal:
		entityTypeStorageMapping:
			node: Drupal\node\NodeStorage
			taxonomy_term: Drupal\taxonomy\TermStorage
			user: Drupal\user\UserStorage
```

Now there is an `entityMapping` property so that we can handle future enhancements with entity type static analysis.

```
parameters:
	drupal:
		entityMapping:
			node:
				class: Drupal\node\Entity\Node
				storage: Drupal\node\NodeStorage
			taxonomy_term:
				class: Drupal\taxonomy\Entity\Term
				storage: Drupal\taxonomy\TermStorage
			user:
				class: Drupal\user\Entity\User
				storage: Drupal\user\UserStorage
			block:
				class: Drupal\block\Entity\Block
```

<hr /><em>This discussion was created from the release <a href='https://github.com/mglaman/phpstan-drupal/releases/tag/1.1.0'>1.1.0</a>.</em>

**Comment by eiriksm** [comment]
OK, so here's a question:

Maybe we should already be thinking about bundle classes? Coming in Drupal 9.3? https://www.drupal.org/node/3191609

That being said. This is, amazing work. Totally awesome!

**Reply by mglaman** [reply]
Good question! I thought about it. I also thought I opened an issue, but I did not. I got caught up on the following problem:

There is a problem with bundle classes: we have no way of knowing a bundle of an entity when it is loaded. I think this will fallback on users properly using `instanceof` checks and `assert()` to verify an entity if of a bundle and its bundle class.

**Reply by eiriksm** [reply]
Sounds good. Real easy fix, and it's probably not like people will not do that (or annotating) and expect phpstan-drupal to know...

**Reply by mglaman** [reply]
What we could do is make sure it returns a union type of the base entity class and any known bundle classes. That'd be better than nothing
